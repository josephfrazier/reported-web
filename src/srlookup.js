import { JSDOM } from 'jsdom';

const axios = require('axios');

export default async function srlookup({ reqnumber }) {
  const url = `https://portal.311.nyc.gov/sr-details/?srnum=${reqnumber}`;

  const { data } = await axios.get(url);
  const { document } = new JSDOM(data).window;

  const result = {};
  result.description = document.querySelector('#page-wrapper p')?.textContent;
  const fields = [...document.querySelectorAll('.info, .control')];
  for (let i = 0; i < fields.length; i += 2) {
    const keyField = fields[i];
    const valueField = fields[i + 1];

    const key = keyField.textContent;
    const value = valueField.textContent;

    result[key] = value;
  }

  const srdatereported = /\$\("#srdatereported"\).text\(getESTDate\("([^"]+)"\)\)/.exec(
    data,
  );
  if (srdatereported) {
    result['Date Reported'] = new Date(
      srdatereported[1],
    ).toLocaleString('en-US', { timeZone: 'America/New_York' });
  }

  const srupdatedon = /\$\("#srupdatedon"\).text\(getESTDate\("([^"]+)"\)\)/.exec(
    data,
  );
  if (srupdatedon) {
    result['Updated On'] = new Date(srupdatedon[1]).toLocaleString('en-US', {
      timeZone: 'America/New_York',
    });
  }

  const srdateclosed = /\$\("#srdateclosed"\).text\(getESTDate\("([^"]+)"\)\)/.exec(
    data,
  );
  if (srdateclosed) {
    result['Date Closed'] = new Date(srdateclosed[1]).toLocaleString('en-US', {
      timeZone: 'America/New_York',
    });
  }

  return result;
}
