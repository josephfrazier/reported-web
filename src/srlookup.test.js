/**
 * @jest-environment node
 */
/* eslint-env jest */

import srlookup from './srlookup.js';

describe('srlookup', () => {
  test('returns the right object', async () => {
    const result = await srlookup({ reqnumber: '311-18685922' });

    expect(result).toMatchSnapshot();
  });
});
