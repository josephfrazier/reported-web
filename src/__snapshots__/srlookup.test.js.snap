// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`srlookup returns the right object 1`] = `
Object {
  "Additional Details": "Unsafe Driving",
  "Date Closed": "2019-09-16T16:02:00Z",
  "Date Reported": "2019-07-03T14:43:28Z",
  "Problem": "For Hire Vehicle Complaint",
  "Problem Details": "Driver Complaint - Non Passenger",
  "SR Address": "",
  "SR Number": "311-00039062",
  "SR Status": "Closed",
  "Time To Next Update": " N/A ",
  "Updated On": "2019-09-16T16:02:22Z",
  "description": "A hearing was scheduled at the OATH/Taxi and Limousine Tribunal.",
}
`;
